tracer/src/Datadog.Trace/TraceContext.cs

```diff
@@ -240,28 +246,40 @@ internal void WriteClosedSpans()
 
         public void SetSamplingPriority(SamplingDecision decision, bool notifyDistributedTracer = true)
         {
-            SetSamplingPriority(decision.Priority, decision.Mechanism, notifyDistributedTracer);
+            SetSamplingPriority(decision.Priority, decision.Mechanism, decision.Rate, decision.LimiterRate, notifyDistributedTracer);
         }
 
-        public void SetSamplingPriority(int? priority, int? mechanism = null, bool notifyDistributedTracer = true)
+        public void SetSamplingPriority(
+            int? priority,
+            string? mechanism = null,
+            double? rate = null,
+            double? limiterRate = null,
+            bool notifyDistributedTracer = true)
         {
             if (priority == null)
             {
                 return;
             }
 
+            // priority (keep/drop) can change (manually, ASM, etc)
             _samplingPriority = priority;
 
-            const string tagName = Trace.Tags.Propagated.DecisionMaker;
+            // report only the original rates, do not override
+            InitialSamplingRate ??= rate;
+            LimiterSamplingRate ??= limiterRate;
 
             if (priority > 0 && mechanism != null)
             {
-                Tags.TryAddTag(tagName, SamplingMechanism.GetTagValue(mechanism.Value));
+                // report sampling mechanism only if decision is to keep the trace.
+                // report only the original sampling mechanism, do not override.
+                InitialSamplingMechanism ??= mechanism;
+                Tags.TryAddTag(Trace.Tags.Propagated.DecisionMaker, mechanism);
             }
             else if (priority <= 0)
             {
-                // remove tag if priority is AUTO_DROP (0) or USER_DROP (-1)
-                Tags.RemoveTag(tagName);
+                // remove sampling mechanism if decision is to drop the trace
+                InitialSamplingMechanism = null;
```

Comment:
TODO FIX: this allows the sampling mechanism to change is a user drop then keeps the trace. Instead we should always report the original mechanism if it was not "manual".